Floating point operations from serde

When running `make debug` I receive the following output:

```
cargo build --package tez_kernel --target wasm32-unknown-unknown --profile release --target-dir ./target/repl
    Finished release [optimized] target(s) in 0.32s
wasm-strip -o ./.bin/debug_kernel.wasm ./target/repl/wasm32-unknown-unknown/release/tez_kernel.wasm
docker run --rm -it --entrypoint=/bin/sh --name wasm-repl -v $PWD/.bin:/.bin tezos/tezos:master_f9675b19_20230303221946 -c '/usr/local/bin/octez-smart-rollup-wasm-debugger /.bin/debug_kernel.wasm --inputs /.bin/inputs.json'
Error
  Unknown exception: Error(_, "float instructions are forbidden")

make: *** [debug] Error 1
```

Running `grep -nE 'f(32|64)\.' ./.bin/kernel.wat` I can find a few occurrences of floating point instructions, and looking into the wat file I could find the following functions, with mangled names:

- `$_ZN77_$LT$serde..__private..de..content..Content$u20$as$u20$core..clone..Clone$GT$5clone17h12cafb7e0f652464E`: Which is the result of the Clone derive macro on https://github.com/serde-rs/serde/blob/master/serde/src/private/de.rs#L219
- `$_ZN5serde9__private2de7content7Content10unexpected17h6a53c07a72eeeb43E`: From https://github.com/serde-rs/serde/blob/master/serde/src/private/de.rs#L274
- `$_ZN60_$LT$serde..de..Unexpected$u20$as$u20$core..fmt..Display$GT$3fmt17hcb6eab2fe1eecf7bE`: From https://github.com/serde-rs/serde/blob/master/serde/src/de/mod.rs#L403
- `$_ZN4core3fmt5float29float_to_decimal_common_exact17hd806da160a15dfd7E`: From https://github.com/rust-lang/rust/blob/master/library/core/src/fmt/float.rs#L29, which probably comes from inlining some of those function calls inside it. I believe this one can be circumvented by having `cfg(no_fp_fmt_parse)` enabled. But probably could optimized  out once float dependency is removed from `Unexpected`.
- `$_ZN4core3fmt5float52_$LT$impl$u20$core..fmt..Display$u20$for$u20$f64$GT$3fmt17hbe7611732fc77d82E`: From https://github.com/rust-lang/rust/blob/master/library/core/src/fmt/float.rs#L225

I'm not sure how to solve this, but I believe we could do something with conditional compilation directly on `serde`, since `cfg(no_integer128)` (example https://github.com/serde-rs/serde/blob/master/serde/src/de/mod.rs#L121) opens this possibility.

--- a/Makefile
+++ b/Makefile
@@ -90,11 +90,13 @@ operator-shell:
 wat:
 	cargo build --package tez_kernel --target wasm32-unknown-unknown
 	wasm2wat -o ./.bin/kernel.wat ./target/wasm32-unknown-unknown/debug/tez_kernel.wasm
+	# check if there's no floating point calc
+	grep -nE 'f(32|64)\.' ./.bin/kernel.wat || true
 
 debug:
 	cargo build --package tez_kernel --target wasm32-unknown-unknown --profile release --target-dir ./target/repl
 	wasm-strip -o ./.bin/debug_kernel.wasm ./target/repl/wasm32-unknown-unknown/release/tez_kernel.wasm
-	docker run --rm -it --entrypoint=/bin/sh --name wasm-repl -v $$PWD/.bin:/root/.bin tezos/tezos:$(MONDAY_TAG) /usr/local/bin/octez-smart-rollup-wasm-debugger /root/.bin/debug_kernel.wasm --inputs /root/.bin/inputs.json
+	docker run --rm -it --entrypoint=/usr/local/bin/octez-smart-rollup-wasm-debugger --name wasm-repl -v $$PWD/.bin:/home/.bin tezos/tezos:$(MONDAY_TAG) /home/.bin/debug_kernel.wasm --inputs /home/.bin/inputs.json
 
 monday:
 	$(MAKE) build-operator

--- a/tezos_ctx/src/types/node.rs
+++ b/tezos_ctx/src/types/node.rs
@@ -89,9 +89,17 @@ context_node_type_core!(Nat);
 macro_rules! context_node_type_rpc {
     ($ty: ty) => {
         impl ContextNodeType for $ty {
-            fn decode(bytes: &[u8]) -> Result<ContextNode> {
-                let value: $ty = serde_json_wasm::de::from_slice(bytes)?;
-                Ok(value.into())
+            fn decode(_bytes: &[u8]) -> Result<ContextNode> {
+                #[cfg(not(target_arch = "wasm32"))]
+                {
+                    // This is a workaround to avoid floating point operations introduced by serde.
+                    // Since we do not need RPC models deserialization inside the kernel,
+                    // we can only enable that for tests and binaries that are not compiled to wasm.
+                    let value: $ty = serde_json_wasm::de::from_slice(_bytes)?;
+                    return Ok(value.into());
+                }
+                #[cfg(target_arch = "wasm32")]
+                unimplemented!()
             }
 
             fn encode(&self) -> Result<Vec<u8>> {
@@ -101,7 +109,6 @@ macro_rules! context_node_type_rpc {
     };
 }
 
-context_node_type_rpc!(Head);
 context_node_type_rpc!(Operation);
 context_node_type_rpc!(BatchReceipt);
 
@@ -119,5 +126,16 @@ impl ContextNodeType for i64 {
 
     fn encode(&self) -> Result<Vec<u8>> {
         Ok(self.to_be_bytes().to_vec())
+    }
+}
+
+impl ContextNodeType for Head {
+    fn decode(bytes: &[u8]) -> Result<ContextNode> {
+        let value: Head = serde_json_wasm::de::from_slice(bytes)?;
+        Ok(value.into())
+    }
+
+    fn encode(&self) -> Result<Vec<u8>> {
+        Ok(serde_json_wasm::ser::to_vec(self)?)
     }
 }

